Feature: share articles (front-end part)

mnassabain opened the pull request:
Part 2 (front-end) of #1191

SMillerDev commented on js/service/ShareResource.js:
```diff
@@ -0,0 +1,36 @@
+
+    ShareResource.prototype.getUsers = function (search) {
+        return this.http({
+            url: OC.linkToOCS(`apps/files_sharing/api/v1/sharees?search=${search}&itemType=file&perPage=5`, 1),
```
What if there are users on the server that don't have access to news, only to files. They would never be able to view what you shared.

mnassabain replied:
That's true, I didn't really think about that..
We use the ocs sharee api, and i don't really see a solution from the [documentation](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/OCS/ocs-sharee-api.html)

mnassabain replied:
Maybe by writing a function specific to news which does something like
```sql
select distinct(user_id) from feeds where upper(user_id) like upper(‘%query%’)
```
?

SMillerDev replied:
Hmm, if that is the documented way I'm fine with it. The item type is wrong in that case though.

NicoWend replied:
There's no other itemType than "file" and "folder". If the type is different, it falls in a else and does a default behavior. 
However, I changed it to news_item it's much cleaner.

SMillerDev commented on templates/part.content.php:
```diff
@@ -88,6 +88,78 @@ class="icon-toggle keep-unread"
+                                        ng-class="{'icon-loading-small': App.loading.isLoading(user.value.shareWith), 'icon-checkmark': !App.loading.isLoading(user.value.shareWith) && Share.usersSharedArticles[item.id].includes(user.value.shareWith)}">
+                                    </span>
+                                </a>
+                                <p class="label-group"> <?php p($l->t('Share on social media')) ?> </p>
+                            <div class="row">
+                                <div class="col-4">
+                                    <a target="_blank"
+                                        class="icon-dropdown icon-facebook pr-5"
+                                        ng-href="https://www.facebook.com/sharer/sharer.php?u={{ ::item.url }}"></a>
+                                </div>
+                                <div class="col-4">
+                                    <a target="_blank"
+                                        class="icon-dropdown icon-twitter pr-5"
+                                        ng-href="https://twitter.com/intent/tweet?url={{ ::item.url }}"></a>
+                                </div>
+                                <div class="col-4">
+                                    <a class="icon-dropdown icon-mail pr-5"
+                                        ng-href="mailto:?subject=I wanted you to see this site&amp;body=Check out this site {{ ::item.url }}"></a>
```
I believe we should only enable  for example mail if the app is enabled: https://apps.nextcloud.com/apps/socialsharing_email

aureliendvd replied:
Hi,
What does this method bring more than just sending an email?
Another question, how to implement this application directly in news? 
Thank you

SMillerDev commented on img/checkmark.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" height="16" width="16" version="1.1" viewBox="0 0 16 16"><path d="m2.35 7.3 4 4l7.3-7.3" stroke="#000" stroke-width="2" fill="none"/></svg>
```
Could these icons be fetched from the sharing apps or the server?

NicoWend replied:
According to [this doc](https://docs.nextcloud.com/server/latest/developer_manual/design/icons.html), it seems that we can fetch any icon from the server instead of having it in an image folder.
For having the "yourdomain" variable from the endpoint, `OC.Backbone.history.location.origin` should do the trick.

anoymouserver replied:
I would prefer to use the available icons provided by the server, if there are suitable ones. The app is currently inconsistent since we also have our own star, link, .. which will presumably be fixed once the new frontend is ready.

NicoWend replied:
I found a much simpler way to fetch icons from the server.
However, since the facebook and twitter logos are not a part of the set of nextcloud icons, I had to keep them.

mnassabain commented:
I've updated our front-end changes to match the new back-end implementation.
The code is mergeable with the back-end branch and sharing is functional.
We'll be watching for your feedback! Thanks

SMillerDev commented on templates/part.content.php:
```diff
@@ -88,6 +88,78 @@ class="icon-toggle keep-unread"
+                                </div>
+                                <div class="col-4">
+                                    <a class="icon-dropdown icon-mail pr-5"                                        
+                                        ng-href="mailto:?subject=I wanted you to see this site&amp;body=Check out this site {{ ::item.url }}"></a>
```
Can we make this field translateble.

SMillerDev commented on templates/part.content.php:
```diff
@@ -88,6 +88,78 @@ class="icon-toggle keep-unread"
+                                        ng-href="https://www.facebook.com/sharer/sharer.php?u={{ ::item.url }}"></a>
+                                </div>
+                                <div class="col-4">
+                                    <a target="_blank"
+                                        class="icon-dropdown icon-twitter pr-5"
+                                        ng-href="https://twitter.com/intent/tweet?url={{ ::item.url }}"></a>
+                                </div>
```
Can the social media icons be made conditional on the existence of the relevant share app?

anoymouserver replied:
Since I just found it here a list of the currently available socialsharing apps: [nextcloud/socialsharing](https://github.com/nextcloud/socialsharing).

Not sure of the implementation there is useful for you here. (e.g. https://github.com/nextcloud/socialsharing/blob/d03ba4f3b9db178051be661b2607ce9c8930f396/socialsharing_email/js/socialsharingemail.js#L24-L32)

mnassabain replied:
Thanks for the link and your help!

We've already found and studied this option. Our idea is to use the OCA.Sharing.ExternalLinkActions (like it's done in the file sharing app), and loop over the actions.
However, we do not have access to OCA.Sharing. We've added the OCA global in the `.jshintrc` file, which gives us access to OCA, however it doesn't contain `Sharing`. We're still looking into it to figure out why, and we've also posted a [question](https://help.nextcloud.com/t/how-to-access-oca-sharing/110799) on the forum just in case we take too long to find the solution.

anoymouserver commented:
Some notes after playing around a bit:
- clicking the sharing button in the compact view breaks the entire layout
![compact view](https://user-images.githubusercontent.com/15173012/111033342-16659a80-8411-11eb-8414-5ab623bc3a3d.png)
- the user selection should use the display name for displaying and searching like e.g. the files app does
![news share dialog](https://user-images.githubusercontent.com/15173012/111033494-da7f0500-8411-11eb-817b-29d3c000f769.png) -> ![files share dialog](https://user-images.githubusercontent.com/15173012/111033522-f682a680-8411-11eb-845a-b346c4a7b563.png)
- successfully sharing an item should close the dialog (currently you have to click the sharing button again to close it)
- if the feed item does not have any author info, the dash (`-`) is probably unnecessary
![shared by without author](https://user-images.githubusercontent.com/15173012/111034076-43677c80-8414-11eb-8a45-feec232c217a.png)
- the username after `shared by` is clickable but doesn't lead you anywhere. Also the users display name should be used here instead.

mnassabain commented:
@anoymouserver thank you for the detailed analysis. We are working on correcting these issues.

However I have two questions:
1. Could you tell me what dimensions/configurations are you using to get the compact view? I cannot seem to reproduce the bug you have.
2. Is it a good idea to close the dialog after sharing? It could be annoying to the user if he wanted to share an item with more people. Maybe a better idea would be to be able to close the dialog by clicking somewhere else on the screen, and not necessarily on the dropdown button?

Thanks

anoymouserver commented:
1. The 'compact view' is a separate view which could be enabled in the 'Settings' menu at the bottom left (2nd checkbox).
   In general I'm using NC 20.0.8 with your two branches merged together.  The display is FullHD, so the viewport is only reduced by the browser and task bar.
2. If it doesn't close after sharing it should at least close if you click somewhere else (as you also suggested), I'm not sure how dropdowns for options are handled generally in NC.
   This whole frontend part will probably also be redesigned and unified once the frontend rewrite is done, but until then it should be as usable as possible and I'm not sure how often it happens that an items is shared multiple time in a row (I'm using NC News on my own server without any other accounts, so I'm not representative)

mnassabain commented:
In that case we will try to fix the existing dropdown, and integrate at least the changes that you've suggested.

One more thing, in order to display "shared by displayname", the display name needs to be stored in the `sharedBy` field? But I think for some future features it's important to keep the user id in `sharedBy`. Should we add a new field to store the sharedBy display name? Or would there be a way to update the Item->jsonSerialize() function to convert the user id into the display name?

anoymouserver commented:
I don't know how you could get the display name from the JS but storing it in the `sharedBy` field is not a good idea. This should always be the actual account name, as it is globally unique and used to identify users everywhere .. and the display name could change any time, so storing it is also not an option. Maybe you could provide it via the controller, but that's only a guess.

The user's current display could be requested in the backend code using
```php
// use OCP\IUserManager;
$user = $this->userManager->get($userId);
$displayName = $user->getDisplayName();
```

mnassabain commented:
All right, thanks for the tip! I'll look into it

mnassabain commented:
### To do
- [x]  Localize e-mail share strings 'Check out this site' et l'autre
- [x]  Show user display name in share dropdown
- [x]  Dialog now closes when user clicks on other part of screen
- [x]  Hide dash if an article is shared and without author (`-`)
- [x]  Make username in "shared by ..." unclickable
- [ ]  "shared by [username]" => "shared by [display name]
- [ ]  Fix compact view display glitch
- [x]  Integrate socialsharing apps: show/hide sharing on social media depending on whether the corresponding socialsharing app is enabled/disabled
- [x]  Cleanup + refactor
- [x]  Handle share item error response - display ❌ when failed

mnassabain commented:
All issues from the previous comments were fixed + some other ones that we've found. We've even sleightly refactored the code to tidy it up.
Let us know what you think! Thanks!

PS. I'd like to compliment you @SMillerDev and @anoymouserver on you reactiveness, your comments have been very helpful and we're very grateful for that!

anoymouserver commented:
It looks quite good, nice job!

One further note .. I would hide the social media section if there are no available share apps.
![share without actived social media plugins](https://user-images.githubusercontent.com/15173012/111881634-aa0b0e00-89b1-11eb-951c-cec1d7edcee9.png)

anoymouserver commented on templates/part.content.php:
```diff
@@ -88,13 +89,90 @@ class="icon-toggle keep-unread"
+                            class="icon-category-installed pr-3"
+                            ng-click="Share.shareItem(item.id, user.value.shareWith)">
+                            {{ user.label }}
+                            <span class="right" style="margin-top: 1.1em; margin-right: 1em"
+                                ng-class="{'icon-loading-small': Share.isLoading(user.value.shareWith), 'icon-checkmark': Share.isStatus(item.id, user.value.shareWith, true), 'icon-close': Share.isStatus(item.id, user.value.shareWith, false)}">
+                            </span>
```
Why do you use a dedicated `right` class just for this single use of `float: right;` but add the margins entirely inline?

mnassabain commented:
Good point, I've fixed it so the share on social media part is hidden when there are no social sharing apps enabled.
Also I've cleaned the css up a bit more

anoymouserver reviewed (approved):
Looks good to me .. frontend works as expected 👍

SMillerDev commented:
And can you rebase this one?

SMillerDev commented:
Thanks everyone who worked on the feature!

mnassabain commented:
Thank you @SMillerDev, @anoymouserver and @Grotax for your help!
Working on this app has been a great experience and we're glad to have contributed to it!
